### PR TITLE
Override equals for IndividualUriValues

### LINKED ISSUE #108: Comparing IndividualUriValues in ModelObject#listsEquivalent may be broken?
While reading code, I noticed [this part of the list comparison](https://github.com/spdx/Spdx-Java-Library/blob/5d9fbc5aed6d244921ae392bb8840066eba08fd7/src/main/java/org/spdx/library/model/ModelObject.java#L942-L943) that seems weird to me. Here is my understanding of what's going on in general:
- we iterate over `list1`
- for each element, we try to find an element of `list2` that is equivalent
- if all elements have a corresponding match, the lists are equivalent, else inequivalent

Here is my issue with the `IndividualUriValue` case. Setup:
- we are in the case where `item instanceOf IndividualUriValue`
- we check whether `list2` contains `new SimpleUriValue((IndividualUriValue)item)`. Let's call this newly created instance `simpleUri`.
BUT:
- this will check whether `list2` contains any `element` satisfying `simpleUri.equals(element)`
- [the equals method on SimpleUriValue](https://github.com/spdx/Spdx-Java-Library/blob/5d9fbc5aed6d244921ae392bb8840066eba08fd7/src/main/java/org/spdx/library/model/SimpleUriValue.java#L94-L100) checks for class equality
- so if `element` is an instance of any class _except_ `SimpleUriValue` that implements `IndividualUriValue` (e.g. `ExternalSpdxElement` or `ReferenceType`), it will automatically be false

In case this is too abstract, I can try to come up with some minimal test cases, but may require some time.

Would love to hear your opinion @goneall to determine whether these are indeed bugs, or I am misunderstanding things. 

edit: Fixed the github links (now they link to specific commit instead of master to avoid getting out of sync when master changes)

The following additional issue was split off into https://github.com/spdx/Spdx-Java-Library/issues/114
> One additional issue I realized: Since we are checking for _equivalence_, not _equality_, we need to check equivalence in both directions. Else the `listsEquivalent` function is not symmetric.
Example: We have `list1 = [element11, element12]` and `list2 = [element21, element22]`, where `element11`, `element12` and `element21` are all equivalent but `element22` is completely different. The current algorithm would determine that `list1` is equivalent to `list2`, but `list2` is not equivalent to `list1`.

### PR DESCRIPTION
**Pull request by goneall**
This will cause any individual URI values to be equal if their URI's are equal.

Fixes #108

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

### CODE REVIEW

**Comment by goneall** [comment]
@nicoweidner - Let me know if this change looks reasonable - I'll squash the commits before merging.

Also - I'm thinking about spinning a release in the next couple of days with the fixes, let me know if there are any issues you think need to be fixed before the release.

**Review by nicoweidner — COMMENTED** [review]
It took me a while to understand how the classes relate. Assuming that any `IndividualUriValue` is wholly defined by its individual URI (and all other properties are irrelevant/unused), I think these changes make sense.

There are some technical cases that make me feel uneasy, though (I am a mathematician, I am _always_ looking for pathological cases):
- There are some cross-combinations where e.g. a `ExternalExtractedLicenseInfo` could be equal to a `ExternalSpdxElement` if their URIs agree. I think this cannot actually happen because of format requirements on their ids. I didn't check the patterns or combinations exhaustively.
- A `IndividualUriValue` can never be equal to a `ModelObject` that is not an instance of `IndividualUriValue` - but a `ModelObject` can be equal to a `IndividualUriValue` if their ids agree (in cases where a class extends both).

The first one might actually be considered fine. If the URI is the defining property, then even if the situation is possible or becomes possible by future changes, I can see it being fine to consider such elements equal.

The second one seems problematic though. It is breaking symmetricity of `equals` (I am all about symmetricity these days :grinning: ), and this scenario is actually possible if one picks somewhat artificial ids. I added the following lines to your first test case, and it passed, accepting the annotation as equal to the license info:
```
ExternalExtractedLicenseInfo eel = new ExternalExtractedLicenseInfo(modelStore, namespace, externalDocId + ":" + id, copyManager, true);
Annotation annotation = new Annotation(modelStore, namespace, "DocumentRef-externalDoc:LicenseRef-ID", copyManager, true);
assertEquals(annotation, eel);
```

I wonder if this could be solved by adding a more precise class check in `ModelObject#equals`, i.e.: Instead of
```
if (!(o instanceof ModelObject)) {
    return false;
}
```
use
```
if (o == null || o.getClass() == this.getClass()) {
	return false;
}
```
Do you see any cases (except `IndividualUriValue`) where instances of `ModelObject` should be considered equal even though they are not instances of the exact same class?

File: src/main/java/org/spdx/library/model/license/SpdxNoAssertionLicense.java
Comment (nicoweidner) [reply src/main/java/org/spdx/library/model/license/SpdxNoAssertionLicense.java:63]: By now I believe it is totally intended, so a `SimpleUriValue` can be equal to the `SpdxNoAssertionLicense`. So this can be resolved.

**Comment by goneall** [comment]
@nicoweidner Thanks for the extensive review.  Below are a few responses:

> There are some cross-combinations where e.g. a `ExternalExtractedLicenseInfo` could be equal to a `ExternalSpdxElement` if their URIs agree.

This cannot happen if the spec is followed since the ID's must have different prefixes.  The code that re-inflates the objects from the individuals uses the unique prefixes to build the correct object, so it can not happen in practice either.

> Do you see any cases (except `IndividualUriValue`) where instances of `ModelObject` should be considered equal even though they are not instances of the exact same class?

I can't think of any specific cases, but I would be nervous about making this assumption.  Similar to `IndividualUriValue`, a `ModelObject` would be considered equal if the ID's are equal and they are in the same model store.

There is a fundamental assumption in the spec (and carried forward in the code) that the ID's are unique for each element/ModelObject.

Rather than changing equals in `ModelObject`, perhaps we should prevent new `ModelObjects` of a different type from being created using the same ID's.  In your example above, the creation of the `Annotation` should fail for 2 reasons:
- The ID uses a `LicenseRef-` prefix which is reserved for license types
- The ID already exists in a different class

If you assume that ID's are unique, would the equals be considered symmetric?

File: src/main/java/org/spdx/library/model/SpdxNoneElement.java
Comment (goneall) [reply src/main/java/org/spdx/library/model/SpdxNoneElement.java:73]: Correct - It also make the hash code algorithm consistent.

File: src/test/java/org/spdx/library/model/IndividualUriValueTest.java
Comment (goneall) [reply src/test/java/org/spdx/library/model/IndividualUriValueTest.java:36]: Since the checksum is a required field, I thought it would be good to add it to the tests as well.

File: src/main/java/org/spdx/library/model/ExternalDocumentRef.java
Comment (goneall) [reply src/main/java/org/spdx/library/model/ExternalDocumentRef.java:239]: Good catch - I believe this is was an incorrect addition.  I'll update the PR removing this check.

Comment (goneall) [reply src/main/java/org/spdx/library/model/ExternalDocumentRef.java:239]: In looking back at the code, the override is for an `IndividualUriValue` created inline for a property value, so I believe the code is correct:

```
			setPropertyValue(SpdxConstants.PROP_EXTERNAL_SPDX_DOCUMENT,
					new IndividualUriValue() {
						@Override ...
```

Comment (nicoweidner) [reply src/main/java/org/spdx/library/model/ExternalDocumentRef.java:239]: Oh, you are totally right, I didn't look close enough :flushed:
But then I wonder why `SimpleUriValue` with `documentNamespace` as uri isn't used here? I assumed that this was meant as a generic implementing class of `IndividualUriValue`, and I believe it would behave the same (apart from a technical difference in `hashCode` that shouldn't matter)

**Comment by nicoweidner** [comment]
> @nicoweidner Thanks for the extensive review. Below are a few responses:
> 
> > There are some cross-combinations where e.g. a `ExternalExtractedLicenseInfo` could be equal to a `ExternalSpdxElement` if their URIs agree.
> 
> This cannot happen if the spec is followed since the ID's must have different prefixes. The code that re-inflates the objects from the individuals uses the unique prefixes to build the correct object, so it can not happen in practice either.

Yeah, I think this case is fine.

> > Do you see any cases (except `IndividualUriValue`) where instances of `ModelObject` should be considered equal even though they are not instances of the exact same class?
> 
> I can't think of any specific cases, but I would be nervous about making this assumption. 

Do you mean you are hesitant to change the `ModelObject#equals` method to require strict class equality? My idea was to put in this requirement for the method on the `ModelObject` level, and then if some subclass wants to use a different implementation, it can override it (as is done for the implementations of `IndividualUriValue`).

> Similar to `IndividualUriValue`, a `ModelObject` would be considered equal if the ID's are equal and they are in the same model store.

Do you consider one model store as basically the same as one document? Because technically, it would check for `documentUri`, not the store, right? (unless it's an anonymous id)
I am asking because while working on some tooling, I dealt with conversion cases, where two stores for the same document can be in play (for two different formats), so they would have the same documentUri/namespace.

> There is a fundamental assumption in the spec (and carried forward in the code) that the ID's are unique for each element/ModelObject.
> 
> Rather than changing equals in `ModelObject`, perhaps we should prevent new `ModelObjects` of a different type from being created using the same ID's. In your example above, the creation of the `Annotation` should fail for 2 reasons:
> 
> * The ID uses a `LicenseRef-` prefix which is reserved for license types
> * The ID already exists in a different class
> 
> If you assume that ID's are unique, would the equals be considered symmetric?

Haha... I started writing a similar comment in response to your "id's are expected to be unique" statement, then realized you had already written basically the same thing.
I believe that should work as well, yes. I would prefer a solution that doesn't even leave technical loopholes, but it seems fine in practice.
The only situation I can think of that could produce weird results would be the conversion scenario mentioned above, where I have two model stores describing the same document, for different formats (say, tag and json). Then my "counterexample" above would still work. It does appear rather far-fetched at this point though, and if the spec requires specific id formats, I'd say it can be ignored (it would be nice if any id format requirements are also enforced, though!).


Offtopic remark: Thanks for being patient in your explanations of the backgrounds! As I'm still missing a lot of context knowledge, being pedantic helps me understand the precise relationships between objects.

**Comment by goneall** [comment]
> Do you mean you are hesitant to change the ModelObject#equals method to require strict class equality? My idea was to put in this requirement for the method on the ModelObject level, and then if some subclass wants to use a different implementation, it can override it (as is done for the implementations of IndividualUriValue).

Yes - I'm a bit hesitant since the namespace + ID's must be unique and if the namespace + ID is the same, then the should be equal - even if they are instantiated in different classes.  BTW - if everything is working properly, than the classes shouldn't be different.  I would just want that namespace + ID to take precedence on equality.

> Do you consider one model store as basically the same as one document? Because technically, it would check for documentUri, not the store, right? (unless it's an anonymous id)
I am asking because while working on some tooling, I dealt with conversion cases, where two stores for the same document can be in play (for two different formats), so they would have the same documentUri/namespace.

Typically, there is one document in one model store, but the design allows for multiple document URI's and multiple documents in the same store.  Most `IModelStore` methods take a `documentUri` as one of the parameters.  You could not, however, have 2 documents with the same URI in the same ModelStore  The `CopyManager` takes care of conversions between the model stores.

> it would be nice if any id format requirements are also enforced, though

This would be my preferred approach.  There is a field `strict` which causes exceptions when an object is created or a parameter set which technically doesn't meet the spec.  I'm thinking we should check the ID format and throw an exception if `strict` is true, otherwise report a validation error.

> Offtopic remark: Thanks for being patient in your explanations of the backgrounds! As I'm still missing a lot of context knowledge, being pedantic helps me understand the precise relationships between objects.

No worries - I'm very appreciative of the review and comments.  It really helps build a better library.

**Comment by goneall** [comment]
@nicoweidner I did some refactoring of the common code for `IndividualUriValue` `equals` and `hashCode`.

Unless you see any major issues, I'll merge this after it passes the CI.  I'll submit another PR that should help with the symmetric equals.

File: src/main/java/org/spdx/library/model/ExternalDocumentRef.java
Comment (nicoweidner) [inline src/main/java/org/spdx/library/model/ExternalDocumentRef.java:237]: Is there a reason for using an anonymous class instead of a `SimpleUriValue` instance now?

File: src/main/java/org/spdx/library/model/SimpleUriValue.java
Comment (nicoweidner) [inline src/main/java/org/spdx/library/model/SimpleUriValue.java:67]: These could just be default methods on `IndividualUriValue`, right?

**Review by nicoweidner — APPROVED** [review]
I only left some small comments, looks good to me now!

File: src/main/java/org/spdx/library/model/ExternalDocumentRef.java
Comment (goneall) [reply src/main/java/org/spdx/library/model/ExternalDocumentRef.java:237]: I'm just using `SimpleUriValue` as a convenient place to refactor the common code for all `IndividualUriValue` `equals` and `hashCode` as static methods.  It would probably be clearer if I created a separate static helper class, but I didn't want to create another class with just those methods.

File: src/main/java/org/spdx/library/model/SimpleUriValue.java
Comment (goneall) [reply src/main/java/org/spdx/library/model/SimpleUriValue.java:67]: Same as above - just refactoring some common code.

**Pull request merged**
